walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,155 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  /**
+   * Gets the available time slot for new meeting given existing meeting events.
+   * @param events Existing meeting events.
+   * @param request Request to set up a new meeting
+   * @return If one or more time slots exists so that both mandatory and optional attendees can attend,
+   *         returns those time slots. Otherwise, returns the time slots that fit just the mandatory attendees.
+   *         Returns empty if duration is greater than a day.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (events.isEmpty()) {
+      if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return new ArrayList<>();
+      } else {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+    }
+
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> attendeesWithOptional = new ArrayList<>();
+    attendeesWithOptional.addAll(request.getAttendees());
+    attendeesWithOptional.addAll(request.getOptionalAttendees());
+
+    // No attendees at all.
+    if (attendeesWithOptional.isEmpty()) {
+      return new ArrayList<>();
+    }
+
+    long duration = request.getDuration();
+
+    List<TimeRange> occupiedTimeRange = new ArrayList<>();
+    List<TimeRange> occupiedTimeRangeWithOptional = new ArrayList<>();
+
+    for (Event event: events) {
+      if (hasAttendeesAttending(event.getAttendees(), attendees)) {
+        occupiedTimeRange.add(event.getWhen());
+      }
+
+      if (hasAttendeesAttending(event.getAttendees(), attendeesWithOptional)) {
+        occupiedTimeRangeWithOptional.add(event.getWhen());
+      }
+    }
+
```

Comment:
Look like all the code above this line is to get two collection of occupied time. Maybe abstract them out to separate method?

Reply:
Could also be
List<TimeRange> occupiedTimeRange = getOccupiedTimeRange(attendees, event);
List<TimeRange> occupiedTimeRangeXXX = getOccupiedTimeRange(attendeesXXX, event);